Offer the ability to pass through standard capnslog when embedding.

purpleidea opened the pull request:
This resolves: https://github.com/coreos/etcd/issues/4115

purpleidea commented:
Here is one patch I've made to work around https://github.com/coreos/etcd/issues/4115 It doesn't affect any of the capnslog functionality anywhere else as it's only an additional formatter which nobody has to use. I think it could be useful for embedding etcd.

It's a bit of a hack, but it helps for now. You can use it with:

	capnslog.SetFormatter(capnslog.NewSimpleFormatter(os.Stderr))

purpleidea commented:
FWIW I added a second patch with an even more elegant formatter for use in embedding etcd. Cheers

jonboulle commented on capnslog/formatters.go:
```diff
@@ -28,6 +29,28 @@ type Formatter interface {
 
+func NewSimpleFormatter(w io.Writer) *SimpleFormatter {
+	return &SimpleFormatter{
+		w: bufio.NewWriter(w),
```
Totally unused?

purpleidea replied:
It's unused in the lib, but useful to someone consuming it.

jonboulle replied:
It's not exported, so no-one can consume it :-)

purpleidea replied:
Why isn't it exported??

jonboulle replied:
> In Go, a name is exported if it begins with a capital letter.
> When importing a package, you can refer only to its exported names. Any "unexported" names are not accessible from outside the package.

purpleidea replied:
This I know, but either I misunderstand where the git comments are pointing to, or I don't see the lack of capitalization... Are we talking about: NewSimpleFormatter ? It looks capitalized to me.

jonboulle replied:
Talking about the field `w` on the `SimpleFormatter` struct - e.g. this won't work:
``
import "cpnslog"
f := capnslog.NewSimpleFormatter(bytes.Buffer{})
fmt.Println(f.w) // can't access w from outside package!
```

jonboulle replied:
and even if it IS exported I don't see what it provides to consumers..

jonboulle commented on capnslog/formatters.go:
```diff
@@ -28,6 +29,28 @@ type Formatter interface {
+	}
+}
+
+type SimpleFormatter struct {
```
I know we've failed at this so far, but docstrings explaining the different formatters would be very helpful for consumers

purpleidea replied:
Strangely enough, I was going to add some, but there weren't any others so I didn't. I'm happy to add them and rebase if you'd like, I'd just rather not add them for everything :)

jonboulle replied:
Fair, even if you just do it for the ones you've added, that'd be helpful.

jonboulle commented on capnslog/formatters.go:
```diff
@@ -104,3 +127,31 @@ func (c *PrettyFormatter) Format(pkg string, l LogLevel, depth int, entries ...i
+}
+
+func (s *LogFormatter) Flush() {
+	s.w.Flush()
```
This doesn't do anything since we never actually write to w right?

purpleidea replied:
The logger object (what the built in log provides) *does* write to w. Not 100% sure the flush is still needed, but I think it fits logically.

jonboulle replied:
These are different `w`s. The logger writes to the passed `io.Writer`; this is calling `Flush()` on a `bufio.Writer` that we wrapped around that `io.Writer`

purpleidea replied:
Yeah, you're totally right. Remind me not to write code at 6am. I'll simplify and rebase and remove the bufio and the flush entirely.

purpleidea commented:
@jonboulle I believe I fixed all the issues, rebased and pushed them back. Annoying that GH looses context on old comments when you force push :P

jonboulle commented on capnslog/formatters.go:
```diff
@@ -104,3 +105,33 @@ func (c *PrettyFormatter) Format(pkg string, l LogLevel, depth int, entries ...i
+}
+
+// Format builds a log message for the LogFormatter. The LogLevel is ignored.
+func (s *LogFormatter) Format(pkg string, l LogLevel, i int, entries ...interface{}) {
```
nit, but can you change the receiver name to "l" or "f" or "lf" instead of "s"? here and for Flush()

jonboulle replied:
Also consider `(pkg string, _ logLevel, i int, entries ...interface{})`

jonboulle commented on capnslog/formatters.go:
```diff
@@ -104,3 +105,33 @@ func (c *PrettyFormatter) Format(pkg string, l LogLevel, depth int, entries ...i
+
+// NewLogFormatter is a helper to produce a new LogFormatter struct. It uses the
+// golang log package to actually do the logging work so that logs look similar.
+func NewLogFormatter(w io.Writer, prefix string, flag int) *LogFormatter {
```
This should just return a `Formatter`, not a LogFormatter

jonboulle replied:
(I don't know why NewStringFormatter doesn't do the same, looks like a bug to me)

jonboulle commented on capnslog/formatters.go:
```diff
@@ -104,3 +105,33 @@ func (c *PrettyFormatter) Format(pkg string, l LogLevel, depth int, entries ...i
+	s.logger.Output(5, fmt.Sprintf("%s%v", prefix, str)) // call depth is 5
+}
+
+// Flush is included so that the interface is complete, but is not used.
```
Rather than "is not used", prefer "is a no-op". It's up to users of the interface to decide whether it's used or not ;-)

jonboulle commented:
Few final comments. Could you please also tweak the first line of your commit message?
	`capnslog: add formatter to simulate normal log package behaviour`

or similar. Thanks!

purpleidea commented:
@jonboulle This was actually a really excellent review (even the nitpick) thanks! Can I get you as a reviewer on https://github.com/purpleidea/mgmt/ ?

I've rebased everything, and also fixed the StringFormatter return value in a separate patch. LMK.

jonboulle commented:
@purpleidea you can try CCing me on some reviews, but I can't promise I'll get to anything in a timely manner :-) (note the two week gap on this one).

Thanks for your patience - this LGTM.

purpleidea commented:
woo!

Mgmt will get more interesting as the etcd parts get more cleaned up and merged :)

Cheers
